orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/monitoreddeploy/MonitoredDeployBaseTask.java

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright 2019 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.orca.clouddriver.tasks.monitoreddeploy;
+
+import com.netflix.spectator.api.Registry;
+import com.netflix.spinnaker.config.DeploymentMonitorDefinition;
+import com.netflix.spinnaker.config.DeploymentMonitorServiceProvider;
+import com.netflix.spinnaker.orca.ExecutionStatus;
+import com.netflix.spinnaker.orca.RetryableTask;
+import com.netflix.spinnaker.orca.TaskResult;
+import com.netflix.spinnaker.orca.clouddriver.pipeline.servergroup.strategies.MonitoredDeployStageData;
+import com.netflix.spinnaker.orca.deploymentmonitor.models.DeploymentStep;
+import com.netflix.spinnaker.orca.deploymentmonitor.models.EvaluateHealthResponse;
+import com.netflix.spinnaker.orca.pipeline.model.Stage;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nonnull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import retrofit.RetrofitError;
+
+public class MonitoredDeployBaseTask implements RetryableTask {
+  private static final int MAX_RETRY_COUNT = 3;
+  protected final Logger log = LoggerFactory.getLogger(getClass());
+  protected DeploymentMonitorDefinition monitorDefinition;
+  protected Stage stage;
+  protected Registry registry;
+
+  private DeploymentMonitorServiceProvider deploymentMonitorServiceProvider;
+
+  MonitoredDeployBaseTask(
+      DeploymentMonitorServiceProvider deploymentMonitorServiceProvider, Registry registry) {
+    this.deploymentMonitorServiceProvider = deploymentMonitorServiceProvider;
+    this.registry = registry;
+  }
+
+  @Override
+  public long getBackoffPeriod() {
+    return TimeUnit.MINUTES.toMillis(1);
+  }
+
+  @Override
+  public long getTimeout() {
+    // TODO(mvulfson): Use DeploymentMonitorDefinition
+    return TimeUnit.MINUTES.toMillis(30);
+  }
+
+  @Override
+  public TaskResult execute(@Nonnull Stage stage) {
+    TaskResult result;
+    MonitoredDeployStageData context = stage.mapTo(MonitoredDeployStageData.class);
+
+    try {
+      this.stage = stage;
+      this.monitorDefinition =
+          deploymentMonitorServiceProvider.getDefinitionById(
+              context.getDeploymentMonitor().getId());
+
+      result = executeInternal();
+    } catch (RetrofitError e) {
```

Comment:
not a big fan of result flowing out of multiple layers of nested if/else in a catch block... I think it reads better if you extract out this catch block in its own method with early returns:

```  private TaskResult handleError(RetrofitError e) {
    registry
      .counter("deploymentMonitor.errors", "monitorId", monitorDefinition.getId())
      .increment();

    int currentRetryCount = context.getDeployMonitorHttpRetryCount();
    if (currentRetryCount < MAX_RETRY_COUNT) {
      // Retry
      return TaskResult.builder(ExecutionStatus.RUNNING)
        .context("deployMonitorHttpRetryCount", ++currentRetryCount)
        .build();
    }
    
    if (monitorDefinition.isFailOnError()) {
      registry
        .counter("deploymentMonitor.fatalErrors", "monitorId", monitorDefinition.getId())
        .increment();

      log.error(
        "Failed to get valid response for {} from deployment monitor {}, aborting because the monitor is marked with failOnError",
        getClass().getSimpleName(),
        monitorDefinition);

      return TaskResult.builder(ExecutionStatus.TERMINAL)
        // TODO(mvulfson)
        // .context()
          .build();
    }
    
    log.warn(
      "Failed to get valid response for {} from deployment monitor {}, ignoring failure",
      getClass().getSimpleName(),
      monitorDefinition);

    return TaskResult.builder(ExecutionStatus.SUCCEEDED)
      // TODO(mvulfson)
      // .context()
        .build();
  }```

Reply:
i am torn... i have a hatred of early/multiple returns. but let me noodle on this

Reply:
i concede that in this case you are right - it does read better :) thanks!